lsコマンドを作る5の提出

maruko92188 opened the pull request.

maruko92188 committed:
lsコマンドを作る5の提出

JunichiIto commented on 04.ls/ls.rb:
```suggestion
file_names = hogehoge(options)
```

メソッド化するとスッキリしそう

maruko92188 replied:
ファイル名を取得する部分を別メソッドにして、mainメソッドの役割を明確にしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -40,26 +40,26 @@ def main
 
-  widths = create_long_format_widths(long_format_table)
-  long_format_table.each do |format|
+  widths = create_long_format_widths(file_status_table)
```
`build_file_status_table`のようにbuildで始まるメソッドもありますが、create/buildの使い分けって何かこだわりがありますか？
（もしくは全体的にcreate/build以外の動詞を検討する、というのもありかも）

maruko92188 replied:
> create/buildの使い分けって何かこだわりがありますか？

`create`を使ったメソッドは戻り値がそのままの状態で使われる、`build`を使ったメソッドは戻り値がその後ループ処理などされる、というイメージで名前をつけていました。

maruko92188 replied:
> （もしくは全体的にcreate/build以外の動詞を検討する、というのもありかも）

「メソッド名は動詞から始まる」というイメージでしたが、
[Railsモデルのメソッドの命名について \#Ruby \- Qiita](https://qiita.com/qpSHiNqp/items/7bcb0492c777a488ceba#%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E5%90%8D)
[モデルやメソッドに名前を付けるときは英語の品詞に気をつけよう \#Ruby \- Qiita](https://qiita.com/jnchito/items/459d58ba652bf4763820#%E6%B3%A8%E6%84%8F%E5%8B%95%E8%A9%9E%E3%81%A7%E5%A7%8B%E3%82%81%E3%82%8B%E3%81%AE%E3%81%AF%E3%81%82%E3%81%8F%E3%81%BE%E3%81%A7%E5%87%A6%E7%90%86%E3%82%92%E5%AE%9F%E8%A1%8C%E3%81%99%E3%82%8B%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8C%E5%AF%BE%E8%B1%A1%E3%81%A7%E3%81%99)
これらの記事を読んだり、[「lsコマンドを作る4」の模範解答](https://bootcamp.fjord.jp/practices/224/submission_answer)をみて、「メソッド名が必ずしも動詞から始まるとは限らない」ということがわかりました。
これらを踏まえて、戻り値を返すだけのメソッドは「戻り値が何かわかる名前」、何かしらの処理を行うメソッドは「何をするかわかる名前」になるようにメソッド名を見直したいと思いますが、いかがでしょうか？
（この方針について間違いや補足がありましたらアドバイスお願いします🙏）

JunichiIto replied:
> createを使ったメソッドは戻り値がそのままの状態で使われる、buildを使ったメソッドは戻り値がその後ループ処理などされる、というイメージで名前をつけていました。

メソッドの使われ方によって動詞を考えるのはちょっとおかしい気がします。
基本的にそのメソッドをどう使うかは、使う側の自由なので。
また、「〜というイメージ」というのもmarukoさん独自ルールなので、僕を含め、他の人にその意図は伝わらないですね。

> これらの記事を読んだり、[「lsコマンドを作る4」の模範解答](https://bootcamp.fjord.jp/practices/224/submission_answer)をみて、「メソッド名が必ずしも動詞から始まるとは限らない」ということがわかりました。
これらを踏まえて、戻り値を返すだけのメソッドは「戻り値が何かわかる名前」、何かしらの処理を行うメソッドは「何をするかわかる名前」になるようにメソッド名を見直したいと思いますが、いかがでしょうか？

ls4の模範解答はいったん忘れてください。
たぶん他のメンターが作ったものなので（注: 今初めて見ました）、僕のレビューに入れ込むと思想が異なるため「混ぜるな危険」になってしまいます。

また、僕が書いた記事に出てきた`full_name`はオブジェクト指向プログラミングの文脈を意図しています。
つまり、

```ruby 
person = Person.new(first_name: 'Junichi', last_name: 'Ito')
person.full_name #=> "Junichi Ito"
```

こんな感じですね。
引数もなく、あたかも属性のひとつに見えるようなメソッドであれば、メソッド名をただの名詞にするのはありです。
逆に言うと、そうでなければ動詞で始めた方が無難です。

あと、前回のコメントの「create/build以外の動詞を検討する」という日本語にも着目してほしいです。
これは「動詞を使わない命名もありですよ」という意図が含まれているのかどうか、日本語ネイティブであれば推測できるんじゃないかと思います（邪推せず、素直に日本語を解釈してください）。

JunichiIto reviewed (commented):
コメントしました〜。

maruko92188 committed:
mainメソッドの修正

ファイル名一覧を取得する部分を別メソッドにした

maruko92188 committed:
メソッド名を修正

create_permissions → determine_permissions
create_last_modified_time → determine_last_modifiled_time
create_long_format_widths → build_widths_table

maruko92188 committed:
パーミッション領域を表す数字を定数化

maruko92188 committed:
変数名を修正

JunichiIto reviewed (commented):
修正ありがとうございます！これでOKです〜 🎉 

参考までに「僕だったらこうするかも」というリファクタリング例を考えてみました。
変更した箇所にコメントを入れているので参考にしてみてください〜。

```ruby 
#! /usr/bin/env ruby
# frozen_string_literal: true

require 'optparse'
require 'etc'

COLUMNS = 3
TAB_WIDTH = 8

ENTRY_TYPES = {
  file: '-',
  blockSpecial: 'b',
  characterSpecial: 'c',
  directory: 'd',
  link: 'l',
  fifo: 'p',
  socket: 's',
  unknown: 'w'
}.freeze

FILE_PERMISSIONS = {
  '0' => '---',
  '1' => '--x',
  '2' => '-w-',
  '3' => '-wx',
  '4' => 'r--',
  '5' => 'r-x',
  '6' => 'rw-',
  '7' => 'rwx'
}.freeze

HALF_A_YEAR_SECONDS = (60 * 60 * 24) * (365.2425 / 2)

def main
  options = ARGV.getopts('alr')
  file_names = search_file_names(options)
  options['l'] ? display_long_format(file_names) : display_column_format(file_names)
end

def search_file_names(options)
  terget = options['a'] ? ['..', '.*', '*'] : ['*']
  file_names = Dir[*terget].sort_by(&:downcase)
  options['r'] ? file_names.reverse : file_names
end

def display_long_format(file_names)
  file_status_table = build_file_status_table(file_names)
  total_blocks = file_status_table.sum { |file_status| file_status[:blocks] }
  puts "total #{total_blocks}"

  widths_table = build_widths_table(file_status_table)
  file_status_table.each do |file_status|
    # 1行分のデータであれば単数形のrowとするか、1行を構成する列の集まりと考えてcolsと呼ぶのがいいと思います
    cols = [
      "#{file_status[:file_mode]} ",
      file_status[:hard_links].rjust(widths_table[:hard_links]),
      "#{file_status[:owner_name].ljust(widths_table[:owner_name])} ",
      "#{file_status[:group_name].ljust(widths_table[:group_name])} ",
      file_status[:byte_size].rjust(widths_table[:byte_size]),
      file_status[:last_modified_time],
      file_status[:path_name]
    ]
    puts cols.join(' ')
  end
end

def build_file_status_table(file_names)
  file_names.map do |file_name|
    status = File.lstat(file_name)
    # file_modeもちょっと長いのでいったん変数に入れるとか
    file_mode = "#{ENTRY_TYPES[status.ftype.to_sym]}#{format_permissions(status)}"
    byte_size = status.blockdev? || status.chardev? ? format('%#x', status.rdev) : status.size.to_s
    path_name = status.symlink? ? "#{file_name} -> #{File.readlink(file_name)}" : file_name
    {
      file_mode:,
      hard_links: status.nlink.to_s,
      owner_name: Etc.getpwuid(status.uid).name,
      group_name: Etc.getgrgid(status.gid).name,
      byte_size:,
      last_modified_time: format_last_modified_time(status),
      path_name:,
      blocks: status.blocks
    }
  end
end

# あるデータを文字列に変換して見た目を整える処理はformat(整形)と呼んであげると良いかも
def format_permissions(status)
  octals = status.mode.to_s(8)[-3..].chars
  special_bits_table = [
    [status.setuid?, 's'],
    [status.setgid?, 's'],
    [status.sticky?, 't']
  ]
  octals.zip(special_bits_table).map do |octal, (is_special, special_symbol)|
    standard_permissions = FILE_PERMISSIONS[octal]
    if is_special
      format_special_permission(standard_permissions, special_symbol)
    else
      standard_permissions
    end
  end.join
end

# ここもformatにしてみた
def format_special_permission(standard_permissions, special_symbol)
  # ちょっと上級テクニックですが、文字列を配列に変換して複数の変数に分割代入してみました
  r, w, x = standard_permissions.chars
  execute = x == 'x' ? special_symbol : special_symbol.upcase
  "#{r}#{w}#{execute}"
end

# ここもformatに
def format_last_modified_time(status)
  differnce = Time.now - status.mtime
  format = differnce > HALF_A_YEAR_SECONDS ? '%_m %e  %Y' : '%_m %e %R'
  status.mtime.strftime(format)
end

def build_widths_table(file_status_table)
  # 同じ処理4回繰り返していたのでto_hメソッドでループ処理しました
  # ループ処理にすれば calculate_max_width をメソッドにしなくてもDRYになります
  %i[hard_links owner_name group_name byte_size].to_h do |key|
    width = file_status_table.map { |file_status| file_status[key].size }.max
    [key, width]
  end
end

def display_column_format(file_names)
  return if file_names.empty?

  build_file_names_table(file_names).each do |row|
    puts row.join.rstrip
  end
end

def build_file_names_table(file_names)
  rows = file_names.size.ceildiv(COLUMNS)
  additional_blanks = COLUMNS * rows - file_names.size
  full_file_names = add_tab_characters(file_names) + Array.new(additional_blanks)
  sliced_file_names = full_file_names.each_slice(rows).to_a
  sliced_file_names.transpose
end

def add_tab_characters(file_names)
  column_width = calculate_column_width(file_names)
  file_names.map do |file_name|
    tab_count = column_width - file_name.size.div(TAB_WIDTH)
    "#{file_name}#{"\t" * tab_count}"
  end
end

def calculate_column_width(file_names)
  longest_file_name = file_names.max_by(&:size).size
  longest_file_name.div(TAB_WIDTH) + 1
end

main
```

maruko92188 committed:
特殊権限を適用する部分で多重代入を使うようにする
